sw/device/tests/sim_dv/keymgr_key_derivation.c

```diff
@@ -0,0 +1,333 @@
+                                    config);
+
+  // Enter low power mode.
+  LOG_INFO("Entering low power");
+  wait_for_interrupt();
+}
+
+/** Key manager configuration steps performed in mask ROM. */
+rom_error_t keymgr_rom_test(void) {
+  ASSERT_OK(keymgr_state_check(kKeymgrStateReset));
+  keymgr_sw_binding_set(&kBindingValueRomExt, &kBindingValueRomExt);
+  keymgr_creator_max_ver_set(kMaxVerRomExt);
+  SEC_MMIO_WRITE_INCREMENT(kKeymgrSecMmioSwBindingSet +
+                           kKeymgrSecMmioCreatorMaxVerSet);
+  sec_mmio_check_values(/*rnd_offset=*/0);
+  sec_mmio_check_counters(/*expected_check_count=*/1);
+  return kErrorOk;
+}
+
+/** Key manager configuration steps performed in ROM_EXT. */
+rom_error_t keymgr_rom_ext_test(void) {
+  ASSERT_OK(keymgr_state_check(kKeymgrStateReset));
+
+  const uint16_t kEntropyReseedInterval = 0x1234;
+  ASSERT_OK(keymgr_init(kEntropyReseedInterval));
+  SEC_MMIO_WRITE_INCREMENT(kKeymgrSecMmioInit);
+  sec_mmio_check_values(/*rnd_offset=*/0);
+
+  keymgr_advance_state();
+  ASSERT_OK(keymgr_state_check(kKeymgrStateInit));
+  LOG_INFO("Keymgr entered Init State");
+
+  keymgr_advance_state();
+  ASSERT_OK(keymgr_state_check(kKeymgrStateCreatorRootKey));
+  LOG_INFO("Keymgr entered CreatorRootKey State");
+
+  // The software binding register lock is reset after advancing the key
+  // manager, so we need to call this function to update sec_mmio expectation
+  // table.
+  keymgr_sw_binding_unlock_wait();
+  sec_mmio_check_values(/*rnd_offset=*/0);
+
+  keymgr_sw_binding_set(&kBindingValueBl0, &kBindingValueBl0);
+  keymgr_owner_int_max_ver_set(kMaxVerBl0);
+  SEC_MMIO_WRITE_INCREMENT(kKeymgrSecMmioSwBindingSet +
+                           kKeymgrSecMmioOwnerIntMaxVerSet);
+  sec_mmio_check_values(/*rnd_offset=*/0);
+
+  keymgr_advance_state();
+  ASSERT_OK(keymgr_state_check(kKeymgrStateOwnerIntermediateKey));
+
+  sec_mmio_check_counters(/*expected_check_count=*/5);
+  return kErrorOk;
+}
+
+bool test_main(void) {
+  rom_error_t result = kErrorOk;
+
+  // This test is expected to run in DEV, PROD or PROD_END states.
+  LOG_INFO("lifecycle state: 0x%x", lifecycle_raw_state_get());
+
+  // Initialize pwrmgr
+  dif_pwrmgr_t pwrmgr;
+  CHECK_DIF_OK(dif_pwrmgr_init(
+      mmio_region_from_addr(TOP_EARLGREY_PWRMGR_AON_BASE_ADDR), &pwrmgr));
+
+  // Initialize aon_timer
+  dif_aon_timer_t aon_timer;
+  CHECK_DIF_OK(dif_aon_timer_init(
+      mmio_region_from_addr(TOP_EARLGREY_AON_TIMER_AON_BASE_ADDR), &aon_timer));
+
+  // Get wakeup reason
+  dif_pwrmgr_wakeup_reason_t wakeup_reason;
+  CHECK_DIF_OK(dif_pwrmgr_wakeup_reason_get(&pwrmgr, &wakeup_reason));
+
+  if (compare_wakeup_reasons(&wakeup_reason, &kWakeUpReasonPor)) {
+    LOG_INFO("Powered up for the first time, program flash");
+
+    // Initialize flash
+    init_flash();
+
+    // Lock otp secret partition
+    lock_otp_secret_partition2();
+
+    // reboot device
+    soft_reboot(&pwrmgr, &aon_timer);
+
+  } else {
```

Comment:
IMO slightly better to check for the correct wakeup reason in this else clause too.

Reply:
Checked the reset cause.